x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -62,7 +61,7 @@ def setup
 
     @download_manager = DownloadManager.new(@metadata)
 
-    initialize_metrics
+    @database_metric.initialize_metrics(@metadata.get_all, @states)
```

Comment:
if `DatabaseManager#setup` is run outside of the agent, can `@database_metric` ever be `nil`?

Your prior implementation defaulted its metric to an instance of `LogStash::Instrument::NamespacedNullMetric` supplied by the method `DatabaseManager#metric`, but now we rely on the ivar directly even though it is not set by `DatabaseManager#initialize`.

Even though the current implementation only sends `DatabaseManager#setup` as a result of a plugin sending `DatabaseManager#subscribe_database_path`, and live plugins only exist in production code after the agent has bootstrapped the `DatabaseManager` with `DatabaseManager#database_metric=`, this still feels a bit precarious to me. If any of those current implementation details were to change, we would be in danger of sending `NilClass#initialize_metrics` here.

I see two options forward for safety:

1. make this call "safe" when `@database_metric` is `nil`, either:
    - with the safe navigation operator (`&.`) every time we reference the `@database_metric` ivar:
      ```suggestion
          @database_metric&.initialize_metrics(@metadata.get_all, @states)
      ```
    - OR with an instance method that provides a null implementation:
      ```suggestion
          database_metric.initialize_metrics(@metadata.get_all, @states)
      ```
      and
      ~~~ ruby
        def database_metric
          @database_metric ||= LogStash::Filters::Geoip::DatabaseMetric.new(nil) # TODO: synchronize with `DatabaseManager#setup`?
        end
      ~~~
2. error with an explicit message if `DatabaseManager#setup` is sent before the singleton database manager has been bootstrapped with a metric

Reply:
`@database_metric` should not be `nil` in the current flow as you said and I agree with your point to make it safer.
I will choose option 1b